src/compas/datastructures/assembly/part.py

```diff
@@ -0,0 +1,187 @@
+from __future__ import print_function
+from __future__ import absolute_import
+from __future__ import division
+
+from collections import deque
+# from functools import reduce
+
+from compas.geometry import Frame
+from compas.geometry import Polyhedron as Shape
+from compas.geometry import Transformation
+# from compas.geometry import multiply_matrices
+from compas.geometry import boolean_union_mesh_mesh
+from compas.geometry import boolean_difference_mesh_mesh
+from compas.geometry import boolean_intersection_mesh_mesh
+
+from ..datastructure import Datastructure
+from ..mesh import Mesh
+
+from .exceptions import FeatureError
+
+
+class Part(Datastructure):
+    """A data structure for representing assembly parts.
+
+    Parameters
+    ----------
+    name : str, optional
+        The name of the part.
+        The name will be stored in :attr:`Part.attributes`.
+    frame : :class:`compas.geometry.Frame`, optional
+        The local coordinate system of the part.
+    shape : :class:`compas.geometry.Shape`, optional
+        The base shape of the part geometry.
+    features : list of tuple(:class:`compas.geometry.Shape`, str), optional
+        The features to be added to the base shape of the part geometry.
+
+    Attributes
+    ----------
+    attributes : dict
+        General object attributes that will be included in the data dict.
+    key : int or str
+        The identifier of the part in the connectivity graph of the parent assembly.
+    frame : :class:`compas.geometry.Frame`
+        The local coordinate system of the part.
+    shape : :class:`compas.geometry.Shape`
+        The base shape of the part geometry.
+    features : list of tuple(:class:`compas.geometry.Shape`, str)
+        The features added to the base shape of the part geometry.
+    transformations : deque of :class:`compas.geometry.Transformation`
+        The stack of transformations applied to the part geometry.
+        The most recent transformation is on the left of the stack.
+        All transformations are with respect to the local coordinate system.
+    """
+
+    operations = {
+        'union': boolean_union_mesh_mesh,
+        'difference': boolean_difference_mesh_mesh,
+        'intersection': boolean_intersection_mesh_mesh
+    }
+
+    @property
+    def DATASCHEMA(self):
+        import schema
+        return schema.Schema({
+            "attributes": dict,
+            "key": int,
+            "frame": Frame,
+            "shape": Shape,
+            "features": list,
+            "transformations": list,
+        })
+
+    @property
+    def JSONSCHEMANAME(self):
+        return 'part'
+
+    def __init__(self, name, frame=None, shape=None, features=None, **kwargs):
+        super(Part, self).__init__()
+        self._frame = None
+        self.attributes = {'name': name or 'Part'}
```

Comment:
should we update `self.attributes` with `kwargs`? (same for `Assembly`) or are assembly/part attributes not supposed to be assigned like that?